src/NetAnalyzers/CSharp/Microsoft.CodeQuality.Analyzers/Maintainability/CSharpUseCrossPlatformIntrinsicsFixer.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Diagnostics;
+using Analyzer.Utilities.Lightup;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeQuality.Analyzers.Maintainability;
+
+namespace Microsoft.CodeQuality.CSharp.Analyzers.Maintainability
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    public sealed class CSharpUseCrossPlatformIntrinsicsFixer : UseCrossPlatformIntrinsicsFixer
+    {
+        protected override SyntaxNode CreateExclusiveOrExpression(SyntaxNode left, SyntaxNode right)
+            => SyntaxFactory.BinaryExpression(SyntaxKind.ExclusiveOrExpression, (ExpressionSyntax)left, (ExpressionSyntax)right);
+
+        protected override SyntaxNode CreateLeftShiftExpression(SyntaxNode left, SyntaxNode right)
+            => SyntaxFactory.BinaryExpression(SyntaxKind.LeftShiftExpression, (ExpressionSyntax)left, (ExpressionSyntax)right);
+
+        protected override SyntaxNode CreateRightShiftExpression(SyntaxNode left, SyntaxNode right)
+            => SyntaxFactory.BinaryExpression(SyntaxKind.RightShiftExpression, (ExpressionSyntax)left, (ExpressionSyntax)right);
+
+        protected override SyntaxNode? CreateUnsignedRightShiftExpression(SyntaxNode left, SyntaxNode right)
+        {
+            const LanguageVersion CSharp11 = (LanguageVersion)1100;
+
+            if (!Enum.IsDefined(typeof(SyntaxKind), SyntaxKindEx.UnsignedRightShiftExpression))
+            {
+                return null;
+            }
+
+            if ((left.SyntaxTree.Options is not CSharpParseOptions csharpParseOptions) || (csharpParseOptions.LanguageVersion < CSharp11))
+            {
+                return null;
+            }
+
+            return SyntaxFactory.BinaryExpression(SyntaxKindEx.UnsignedRightShiftExpression, (ExpressionSyntax)left, (ExpressionSyntax)right);
+        }
+
+        protected override SyntaxNode ReplaceWithBinaryOperator(SyntaxNode currentNode, bool isCommutative, Func<SyntaxNode, SyntaxNode, SyntaxNode?> binaryOpFunc)
+        {
+            if (currentNode is not InvocationExpressionSyntax invocationExpression)
+            {
+                Debug.Fail($"Found unexpected node kind: {currentNode.RawKind}");
+                return currentNode;
+            }
+
+            SeparatedSyntaxList<ArgumentSyntax> arguments = invocationExpression.ArgumentList.Arguments;
+
+            if (arguments.Count != 2)
+            {
+                Debug.Fail($"Found unexpected number of arguments for binary operator replacement: {arguments.Count}");
+                return currentNode;
+            }
```

Comment:
this is fine.  but my STRONG recomendation is: make a helper that does this.  This helper should be called from your analyzer, and you bail out in this case.

Then, you can optionally call the helper from your fixer, asserting that it should suceed (but bailing out gracefully if it doesn't).

Note: Debug asserts practically catch nothing.  Thajt's because all shipping products (and references people have) to our analyzers are always in release mode.

Reply:
This pattern si what many/most actual roslyn analyzers/fixers do.  It is entirely fine to have either a 'Helpers' class that has these static helping function that both znalyzer/fixer use.  Or just expose public static helping functions from teh analyzer that the fixer can use.

Now you write the logic once, and easily use from both.

Reply:
The analyzer is checking the `IInvocationOperation` since that's what it has and not reporting a diagnostic in the case of failure.

The fixer is doing an assert of the `SyntaxNode`, since that's what it has and the analyzer shouldn't be passing through anything incorrectly (it's just breadcrumbs that assist debugging if a future refactoring *does* break an expected invariant). Using the operation on the analyzer side makes a number of checks much easier and so this minimal bit of duplication seems reasonable, especially since I don't think we'd be in a better position if we had the analyzer checking the syntax instead of the operation node.